internal/hammer/loadtest/workers.go

```diff
@@ -231,3 +273,178 @@ func (w *LogWriter) Kill() {
+	return &MMDVerifier{
+		tracker:     tracker,
+		errChan:     errChan,
+		leafMMDChan: leafMMDChan,
+		mmdDuration: mmdDuration,
+	}
+}
+
+// MMDVerifier verifies the signed timestamp against the MMD policy for newly
+// added entries by performing inclusion proof.
+type MMDVerifier struct {
+	tracker     *client.LogStateTracker
+	errChan     chan<- error
+	leafMMDChan chan LeafMMD
+	mmdDuration time.Duration
+	cancel      func()
+}
+
+// Run runs the MMD verifier. This should be called in a goroutine.
+func (v *MMDVerifier) Run(ctx context.Context) {
+	if v.cancel != nil {
+		panic("MMDVerifier was ran multiple times")
+	}
+	ctx, v.cancel = context.WithCancel(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		var leafMMD *LeafMMD
+		for {
+			if leafMMD == nil {
+				mmd, ok := <-v.leafMMDChan
+				if !ok {
+					break
+				}
+				leafMMD = &mmd
+			}
+
+			// Retry if the leaf is not yet integrated into the log.
+			if leafMMD.index >= v.tracker.LatestConsistent.Size {
+				// Verify MMD timestamp.
+				sec := int64(leafMMD.timestamp / 1000)
+				nsec := int64((leafMMD.timestamp % 1000) * 1000000)
+				if time.Unix(sec, nsec).Add(v.mmdDuration).Before(time.Now()) {
+					v.errChan <- fmt.Errorf("leaf index %d MMD violation at %d", leafMMD.index, leafMMD.timestamp)
+				}
+
+				break
+			}
+
+			// A new proof builder is required because v.tracker.ProofBuilder
+			// is not thread safe.
+			size := v.tracker.LatestConsistent.Size
+			hash := v.tracker.LatestConsistent.Hash
+			pb, err := client.NewProofBuilder(ctx, log.Checkpoint{
+				Origin: v.tracker.Origin,
+				Size:   size,
+				Hash:   hash,
+			}, v.tracker.TileFetcher)
+			if err != nil {
+				v.errChan <- fmt.Errorf("failed to create proof builder: %w", err)
+				continue
+			}
+			ip, err := pb.InclusionProof(ctx, leafMMD.index)
+			if err != nil {
+				v.errChan <- fmt.Errorf("failed to create inclusion proof: %w", err)
+				continue
+			}
+
+			certs, err := x509.ParseCertificates(leafMMD.leaf)
+			if err != nil {
+				v.errChan <- fmt.Errorf("failed to parse certificates: %w", err)
+				continue
+			}
+			entry, err := entryFromChain(certs, false, leafMMD.timestamp)
+			if err != nil {
+				v.errChan <- fmt.Errorf("failed to create entry from chain: %w", err)
+				continue
+			}
+			leafHash := entry.MerkleLeafHash(leafMMD.index)
+			if err := proof.VerifyInclusion(rfc6962.DefaultHasher, leafMMD.index, size, leafHash, ip, hash); err != nil {
+				v.errChan <- fmt.Errorf("failed to verify inclusion proof: %w", err)
+				continue
+			}
+
+			leafMMD = nil
+		}
+	}
+}
+
+// Kills this verifier at the next opportune moment.
+// This function may return before the verifier is dead.
+func (v *MMDVerifier) Kill() {
+	if v.cancel != nil {
+		v.cancel()
+	}
+}
+
+// entryFromChain generates an Entry from a chain and timestamp.
+// copied from certificate-transparency-go/serialization.go
```

Comment:
Just to keep track of things, can you add a comment / TODO to reflect that it comes form handlers.go?

Reply:
Added the comment.